File: C_Sharp/Treap/TreapNode.cs
using System;
using Monoid;

namespace Treap
{
    [Serializable]
    public class TreapNode<T> : BaseTreapNode<T>
    {
        private T aggregate;
        private readonly IMonoid<T> monoid;

        public TreapNode(T value, IMonoid<T> monoid) : base(value)
        {
            aggregate = value;
            this.monoid = monoid;
        }

        public T Aggregate
        {
            get { return aggregate; }
        }

        protected override void Update()
        {
            base.Update();
            UpdateAggregate();
        }

        public override void UpdateAggregate()
        {
            aggregate = Value;
            if (Left != null)
            {
                aggregate = monoid.Operation(aggregate, ((TreapNode<T>)Left).Aggregate);
            }

            if (Right != null)
            {
                aggregate = monoid.Operation(aggregate, ((TreapNode<T>)Right).Aggregate);
            }
        }

        public override BaseTreapNode<T> RotateLeft()
        {
            BaseTreapNode<T> node = base.RotateLeft();
            TreapNode<T> treapNode = node as TreapNode<T>;
            if (treapNode != null)
            {
                treapNode.UpdateAggregate();
            }

            return node;
        }

        public override BaseTreapNode<T> RotateRight()
        {
            BaseTreapNode<T> node = base.RotateRight();
            TreapNode<T> treapNode = node as TreapNode<T>;
            if (treapNode != null)
            {
                treapNode.UpdateAggregate();
            }

            return node;
        }
    }
}

File: cpp/sa/TreapNode.cpp
#include <stddef.h>
#include "BaseTreapNode.h"
#include "AbstractMonoid.h"
#include "TreapNode.h"

TreapNode::TreapNode(int value) : BaseTreapNode(value)
{
  aggregate = value;
}

int TreapNode::get_aggregate()
{
  return aggregate;
}

void TreapNode::updateAggregate()
{
  aggregate = this->get_value();
  if (this->get_left() != NULL)
    {
      aggregate = operation(aggregate, ((TreapNode*)this->get_left())->get_aggregate());
    }
  if (this->get_right() != NULL)
    {
      aggregate = operation(aggregate, ((TreapNode*)this->get_right())->get_aggregate());
    }
}

BaseTreapNode* TreapNode::rotateLeft()
{
  BaseTreapNode* node = BaseTreapNode::rotateLeft();
  if (node != NULL)
    {
      node->updateAggregate();
    }
  
  return node;
}

BaseTreapNode* TreapNode::rotateRight()
{
  BaseTreapNode* node = BaseTreapNode::rotateRight();
  if (node != NULL)
    {
      node->updateAggregate();
    }
  
  return node;
}

void TreapNode::update()
{
  BaseTreapNode::update();
  updateAggregate();
}
